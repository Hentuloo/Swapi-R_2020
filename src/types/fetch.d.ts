declare function fetch<T>(...args: any): Promise<TypedResponse<T>>;
