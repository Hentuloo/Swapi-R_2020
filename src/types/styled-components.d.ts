import theme from 'themes/mainTheme';

type ThemeInterface = typeof theme;

declare module 'styled-components' {
    // eslint-disable-next-line
    interface DefaultTheme extends ThemeInterface {}
}
