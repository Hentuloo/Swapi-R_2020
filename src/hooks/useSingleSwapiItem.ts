import {
    getSwapiPlanet,
    getSwapiSpecie,
    getSwapiCharacter,
    getSwapiVehicle,
} from 'config/helpers';
import { useQuery } from 'react-query';
import { queryKeys } from 'config/Constants';

export interface UseSingleSwapiItem {
    planetId?: number;
    specieId?: number;
    characterId?: number;
    vehicleId?: number;
}

export const useSingleSwapiItem = ({
    planetId,
    specieId,
    characterId,
    vehicleId,
}: UseSingleSwapiItem) => {
    const character = useQuery({
        queryKey: !!characterId && queryKeys.single.character(characterId),
        queryFn: () => getSwapiCharacter(characterId || 0),
    });

    const planet = useQuery({
        queryKey: !!planetId && queryKeys.single.planet(planetId),
        queryFn: () => getSwapiPlanet(planetId || 0),
    });

    const specie = useQuery({
        queryKey: !!specieId && queryKeys.single.specie(specieId),
        queryFn: () => getSwapiSpecie(specieId || 0),
    });

    const vehicle = useQuery({
        queryKey: !!vehicleId && queryKeys.single.vehicle(vehicleId),
        queryFn: () => getSwapiVehicle(vehicleId || 0),
    });

    return {
        planet,
        specie,
        character,
        vehicle,
    };
};
