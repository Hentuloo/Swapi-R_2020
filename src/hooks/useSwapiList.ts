import { SwapiList, SwapiItemType } from 'types/swapi';
import {
    queryCache,
    usePaginatedQuery,
    PaginatedQueryResult,
    BaseQueryOptions,
} from 'react-query';
import { useState, useCallback } from 'react';

export interface UseSwapiListProps<
    Item,
    Response extends Promise<SwapiList<Item>> = Promise<SwapiList<Item>>
> {
    queryFunc: (page: number) => Response;
    queryKey: (page: number) => string;
    generateQueryKeyForEachItem?: (index: number, item: Item) => string;
    initialPage?: number;
    opts?: BaseQueryOptions;
}
export interface PaginationInterface {
    prevPage: () => void;
    nextPage: () => void;
    activePage: number;
}
export type UseSwapiListResponse<Item> = [
    PaginatedQueryResult<SwapiList<Item>>,
    PaginationInterface,
];

export const useSwapiList = <
    Item extends SwapiItemType,
    QueryResponse extends Promise<SwapiList<Item>> = Promise<SwapiList<Item>>
>({
    queryFunc,
    queryKey,
    generateQueryKeyForEachItem,
    initialPage,
    opts = { suspense: true },
}: UseSwapiListProps<Item, QueryResponse>): UseSwapiListResponse<Item> => {
    const [activePage, setPage] = useState(initialPage || 1);
    const queryResponse = usePaginatedQuery({
        queryKey: queryKey(activePage),
        queryFn: () => queryFunc(activePage),
        config: {
            onSuccess: (data) => {
                //cache fetched list
                if (!generateQueryKeyForEachItem) return;
                data.results.forEach((item, index) => {
                    queryCache.prefetchQuery(
                        generateQueryKeyForEachItem(index, item),
                        () => Promise.resolve(item),
                    );
                });
            },
            ...opts,
        },
    });
    const prevPage = useCallback(() => {
        const { latestData } = queryResponse;
        setPage((old) => (!latestData || !latestData.previous ? old : old - 1));
    }, [queryResponse]);

    const nextPage = useCallback(() => {
        const { latestData } = queryResponse;
        setPage((old) => (!latestData || !latestData.next ? old : old + 1));
    }, [queryResponse]);

    return [queryResponse, { prevPage, nextPage, activePage }];
};
