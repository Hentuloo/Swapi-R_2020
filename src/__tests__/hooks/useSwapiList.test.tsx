import { useSwapiList } from 'hooks/useSwapiList';
import { SwapiCharacter } from 'types/swapi';
import { renderHook, act } from '@testing-library/react-hooks';
import { exampleSwapiCharacterResponse } from '../../tests/exampleSwapiCharacterResponse';
import { queryCache } from 'react-query';

beforeEach(queryCache.clear);

const render = (props?: any) =>
    renderHook(() =>
        useSwapiList<SwapiCharacter>({
            queryFunc: (page) =>
                Promise.resolve(exampleSwapiCharacterResponse[page - 1]),
            queryKey: (page) => 'promise' + (page - 1),
            opts: {
                suspense: false,
            },
            ...props,
        }),
    );

const renderWithData = async (props?: any) => {
    const renderHookResult = render(props);

    //wait query is successful
    await renderHookResult.wait(
        () => renderHookResult.result.current[0].status === 'success',
        { timeout: 1000 },
    );

    return renderHookResult;
};

describe('useSwapiList hook', () => {
    it('snapshot', async () => {
        const { result, unmount } = render();
        expect(result).toMatchSnapshot();

        unmount();
    });
    it('should return data (onSuccess)', async () => {
        const { result, waitForNextUpdate, unmount } = render();

        expect(result.current[0].status).toBe('loading');
        await waitForNextUpdate();
        expect(result.current[0].status).toBe('success');

        const data = result.current[0].resolvedData;
        expect(data).not.toBe(undefined);
        if (!data) return;

        expect(data.results[0].name).toBe(
            exampleSwapiCharacterResponse[0].results[0].name,
        );

        unmount(); // necessary to avoid some 'act' warnings
    });
    it('should deliver paginated results', () =>
        act(async () => {
            const {
                result,
                unmount,
                waitForNextUpdate,
            } = await renderWithData();

            //load first result
            const data = result.current[0].resolvedData;
            expect(data).not.toBe(undefined);
            if (!data) return;
            expect(data.results[0].name).toBe(
                exampleSwapiCharacterResponse[0].results[0].name,
            );

            // change page and wait for update
            expect(result.current[1].activePage).toBe(1);
            result.current[1].nextPage();
            expect(result.current[1].activePage).toBe(2);
            await waitForNextUpdate();

            //check second result
            const secondResult = result.current[0].resolvedData;
            expect(secondResult).not.toBe(undefined);
            if (!secondResult) return;
            expect(secondResult.results[0].name).toBe(
                exampleSwapiCharacterResponse[1].results[0].name,
            );

            unmount(); // necessary to avoid some 'act' warnings
        }));
    it('should prevent fetching next results when it is last page', () =>
        act(async () => {
            const {
                result,
                waitForNextUpdate,
                unmount,
            } = await renderWithData();

            // fetch next page (second page)
            expect(result.current[1].activePage).toBe(1);
            result.current[1].nextPage();
            expect(result.current[1].activePage).toBe(2);
            await waitForNextUpdate();

            // try fetch non-existent page (third page)
            expect(result.current[1].activePage).toBe(2);
            result.current[1].nextPage();
            expect(result.current[1].activePage).toBe(2);

            unmount(); // necessary to avoid some 'act' warnings
        }));
});
