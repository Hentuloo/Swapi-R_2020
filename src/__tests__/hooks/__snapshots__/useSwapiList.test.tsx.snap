// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useSwapiList hook snapshot 1`] = `
Object {
  "current": Array [
    Object {
      "canFetchMore": false,
      "config": Object {
        "cacheTime": 300000,
        "isDataEqual": [Function],
        "onError": [Function],
        "onMutate": [Function],
        "onSettled": [Function],
        "onSuccess": [Function],
        "queryFnParamsFilter": [Function],
        "queryKeySerializerFn": [Function],
        "refetchAllOnWindowFocus": true,
        "refetchInterval": false,
        "refetchOnMount": true,
        "retry": 3,
        "retryDelay": [Function],
        "staleTime": 0,
        "suspense": false,
        "throwOnError": false,
        "useErrorBoundary": undefined,
      },
      "data": undefined,
      "error": null,
      "failureCount": 0,
      "isFetching": true,
      "isStale": true,
      "latestData": undefined,
      "markedForGarbageCollection": false,
      "query": Object {
        "cache": Object {
          "_buildQuery": [Function],
          "cancelQueries": [Function],
          "clear": [Function],
          "getQueries": [Function],
          "getQuery": [Function],
          "getQueryData": [Function],
          "isFetching": 1,
          "prefetchQuery": [Function],
          "queries": Object {
            "[\\"promise0\\"]": [Circular],
          },
          "refetchQueries": [Function],
          "removeQueries": [Function],
          "setQueryData": [Function],
          "subscribe": [Function],
        },
        "cancel": [Function],
        "cancelPromises": [Function],
        "cancelled": null,
        "clear": [Function],
        "config": Object {
          "cacheTime": 300000,
          "isDataEqual": [Function],
          "onError": [Function],
          "onMutate": [Function],
          "onSettled": [Function],
          "onSuccess": [Function],
          "queryFnParamsFilter": [Function],
          "queryKeySerializerFn": [Function],
          "refetchAllOnWindowFocus": true,
          "refetchInterval": false,
          "refetchOnMount": true,
          "retry": 3,
          "retryDelay": [Function],
          "staleTime": 0,
          "suspense": false,
          "throwOnError": false,
          "useErrorBoundary": undefined,
        },
        "fetch": [Function],
        "heal": [Function],
        "instances": Array [
          Object {
            "id": 0,
            "onError": [Function],
            "onSettled": [Function],
            "onStateUpdate": [Function],
            "onSuccess": [Function],
          },
        ],
        "promise": Promise {},
        "queryFn": [Function],
        "queryHash": "[\\"promise0\\"]",
        "queryKey": Array [
          "promise0",
        ],
        "queryVariables": Array [],
        "scheduleGarbageCollection": [Function],
        "scheduleStaleTimeout": [Function],
        "setData": [Function],
        "setState": [Function],
        "state": Object {
          "canFetchMore": false,
          "data": undefined,
          "error": null,
          "failureCount": 0,
          "isFetching": true,
          "isStale": true,
          "markedForGarbageCollection": false,
          "status": "loading",
          "updatedAt": 0,
        },
        "subscribe": [Function],
        "suspenseInstance": Object {
          "onError": [Function],
          "onSettled": [Function],
          "onSuccess": [Function],
        },
        "updateInstance": [Function],
        "wasPrefetched": false,
        "wasSuspended": false,
      },
      "refetch": [Function],
      "resolvedData": undefined,
      "status": "loading",
      "updatedAt": 0,
    },
    Object {
      "activePage": 1,
      "nextPage": [Function],
      "prevPage": [Function],
    },
  ],
  "error": undefined,
}
`;
